sdk/typespec/typespec_client_core/src/http/response_future.rs

```diff
@@ -0,0 +1,520 @@
+use std::{
+    future::{Future, IntoFuture},
+    pin::Pin,
+};
+
+use serde::de::DeserializeOwned;
+use typespec::Error;
+
+use crate::http::{headers::Headers, Response, ResponseBody, StatusCode};
+
+// Below are a series of types that serve to just keep the amount of repetition down.
+// Wasm32 targets don't have Send or Sync, because they are single-threaded, so we need to exclude those from our trait objects on those platforms.
+
+/// Utility shorthand for a Boxed future that outputs T and can capture values with lifetime 'a.
+#[cfg(not(target_arch = "wasm32"))]
+type BoxedFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;
+
+/// Utility shorthand for a Boxed future that outputs T and can capture values with lifetime 'a.
+#[cfg(target_arch = "wasm32")]
+type BoxedFuture<'a, T> = Pin<Box<dyn Future<Output = T> + 'a>>;
+
+/// Utility shorthand for a Boxed Future that takes a ResponseBody and "deserializes" the body, returning a BoxedFuture with the result.
+#[cfg(not(target_arch = "wasm32"))]
+pub type Deserializer<'a, T> =
+    Box<dyn FnOnce(ResponseBody) -> BoxedFuture<'a, Result<T, Error>> + Send + 'a>;
+
+/// Utility shorthand for a Boxed Future that takes a ResponseBody and "deserializes" the body, returning a BoxedFuture with the result.
+#[cfg(target_arch = "wasm32")]
+pub type Deserializer<'a, T> =
+    Box<dyn FnOnce(ResponseBody) -> BoxedFuture<'a, Result<T, Error>> + 'a>;
+
+/// Represents a response where the body has not been downloaded yet, but can be downloaded and deserialized into a value of type `T`.
+///
+/// You get a [`LazyResponse`] by calling [`ResponseFuture::lazy()`] on a [`ResponseFuture`] you received from a service API.
```

Comment:
I'm being utterly pedantic here, but there's something about the term "Lazy" that is bothering me because it feels somewhat pejorative. 

My problem is that I'm not coming up with a better description of how to name it. Lazy has the advantage that it is succinct, whereas other descriptive terms (two step result?) are just way too wordy.

Reply:
If it would help put you at ease, Lazy is used throughout the language
https://doc.rust-lang.org/std/cell/struct.LazyCell.html

Reply:
Yeah, I largely agree @LarryOsterman , but it's definitely become a fairly industry-standard term for "A thing that isn't quite there yet". `DeferredResponse` could be another option, but I feel like `lazy()`/`LazyResponse<T>` is going to be the "expected" name, especially given @RickWinter 's point about its existing use in Rust.

Reply:
I'd rather we stick with `Lazy` as well. It's all over not only in Rust but other languages as well. That, I'm just to lazy to think of a better term.

Reply:
So be it. I'll put my pedantry back in its box.